controllers/effective_ratelimitpolicies_reconciler.go

```diff
@@ -0,0 +1,92 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"sync"
+
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+	"github.com/samber/lo"
+	"k8s.io/client-go/dynamic"
+
+	kuadrantv1 "github.com/kuadrant/kuadrant-operator/api/v1"
+	kuadrantv1beta3 "github.com/kuadrant/kuadrant-operator/api/v1beta3"
+)
+
+type EffectiveRateLimitPolicy struct {
+	Path []machinery.Targetable
+	Spec kuadrantv1beta3.RateLimitPolicy
+}
+
+type EffectiveRateLimitPolicies map[string]EffectiveRateLimitPolicy
+
+type effectiveRateLimitPolicyReconciler struct {
+	client *dynamic.DynamicClient
+}
+
+func (r *effectiveRateLimitPolicyReconciler) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.Reconcile,
+		Events:        rateLimitEventMatchers,
+	}
+}
+
+func (r *effectiveRateLimitPolicyReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("effectiveRateLimitPolicyReconciler")
+
+	kuadrant, err := GetKuadrantFromTopology(topology)
+	if err != nil {
+		if errors.Is(err, ErrMissingKuadrant) {
+			logger.V(1).Info(err.Error())
+			return nil
+		}
+		return err
+	}
+
+	effectivePolicies := r.calculateEffectivePolicies(ctx, topology, kuadrant, state)
+
+	state.Store(StateEffectiveRateLimitPolicies, effectivePolicies)
+
+	return nil
+}
+
+func (r *effectiveRateLimitPolicyReconciler) calculateEffectivePolicies(ctx context.Context, topology *machinery.Topology, kuadrant machinery.Object, state *sync.Map) EffectiveRateLimitPolicies {
+	logger := controller.LoggerFromContext(ctx).WithName("effectiveRateLimitPolicyReconciler").WithName("calculateEffectivePolicies")
+
+	targetables := topology.Targetables()
+	gatewayClasses := targetables.Children(kuadrant) // assumes only and all valid gateway classes are linked to kuadrant in the topology
+	httpRouteRules := targetables.Items(func(o machinery.Object) bool {
+		_, ok := o.(*machinery.HTTPRouteRule)
+		return ok
+	})
+
+	logger.V(1).Info("calculating effective rate limit policies", "httpRouteRules", len(httpRouteRules))
+
+	effectivePolicies := EffectiveRateLimitPolicies{}
+
+	for _, gatewayClass := range gatewayClasses {
+		for _, httpRouteRule := range httpRouteRules {
+			paths := targetables.Paths(gatewayClass, httpRouteRule) // this may be expensive in clusters with many gateway classes - an alternative is to deep search the topology for httprouterules from each gatewayclass, keeping record of the paths
```

Comment:
You are right. That would be simpler. No need to iterate over the gateway classes (as this would be abstracted inside of calculating the topological paths to the http route rules).